### PR TITLE
SSLUsername directive breaks FakeBasicAuth

### PR DESCRIPTION
**Pull request by llabrat**
### Environment
RHEL 6.7
Apache httpd 2.2
Pulp 2.9.2

### Certificate response from /pulp/v2/actions/login/ and pulp-admin login contains ':' in Subject CN:
```
# openssl x509 -inform pem -in user-cert.pem -noout -text
Certificate:
    ... blah blah blah ...
        Subject: CN=admin:admin:57c58e2004c06d1af2f27a8f
    ... blah blah blah ...
```
### Failing request:
```
# /usr/bin/curl -s -k --cert /root/.pulp/user-cert.pem https://vagrant/pulp/api/v2/repositories/
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>403 Forbidden</title>
</head><body>
<h1>Forbidden</h1>
<p>You don't have permission to access /pulp/api/v2/repositories/
on this server.</p>
</body></html>
```
### Apache logs:
```
10.0.2.15 - - [31/Aug/2016:13:55:44 -0400] "GET /pulp/api/v2/repositories/ HTTP/1.1" 403 227 "-" "curl/7.19.7 (x86_64-redhat-linux-gnu) libcurl/7.19.7 NSS/3.18 Basic ECC zlib/1.2.3 libidn/1.18 libssh2/1.4.2"

[Wed Aug 31 11:14:44 2016] [error] [client 10.0.2.15] access to /usr/share/pulp/wsgi/webservices.wsgi failed, reason: FakeBasicAuth is configured and colon (":") character exists in the "admin:admin:57c58e2004c06d1af2f27a8f" username
```
### Passing request after removing SSLUsername directive and restarting httpd:
```
# /usr/bin/curl -s -k --cert /root/.pulp/user-cert.pem https://vagrant/pulp/api/v2/repositories/
[{"scratchpad": {}, "display_name": null, "description": "test", "last_unit_added": null, "notes": {"_repo-type": "rpm-repo"}, "last_unit_removed": null, "content_unit_counts": {}, "_ns": "repos", "_id": {"$oid": "57c70bff04c06d2314d77fc0"}, "id": "test_repo", "_href": "/pulp/api/v2/repositories/test_repo/"}]
```
According to apache documentation [1](https://httpd.apache.org/docs/2.2/mod/mod_ssl.html#sslusername), SSLUsername and FakeBasicAuth do not work in conjunction with one another. After removing SSLUsername directive from vhost, the API once again functions as expected when using certificate for authentication.

1: https://httpd.apache.org/docs/2.2/mod/mod_ssl.html#sslusername

Fixes GH-138

### CODE REVIEW

**Comment by llabrat** [comment]
This change was introduced with [PR122] (https://github.com/Katello/puppet-pulp/pull/112)/[Feature #13189] (http://projects.theforeman.org/issues/13189). There aren't a lot of details about what use case this change was addressing so I cannot confirm this does not break something else.

**Comment by ehelms** [comment]
@llabrat The change was added to allow using a custom derived certificate to used to talk to Pulp via the API. We wanted to be able to create long lived certificates that could communicate to a remote Pulp and this seemed to be the only thing that got it working given Pulp's kind of crazy certificate expectations. I'll do some more testing to see if we can work around this otherwise we could juts make this configurable.

**Comment by llabrat** [comment]
@ehelms I will rewrite to make configurable as soon as I think of a good parameter name. If you can point me to an example of the long lived certificate being created I will play around with that as well.

File: templates/etc/httpd/conf.d/_ssl_vhost.conf.erb [hunk 1]
```diff
@@ -1,7 +1,9 @@
...
-SSLInsecureRenegotiation On
+  SSLInsecureRenegotiation On
 
-KeepAlive On
-MaxKeepAliveRequests <%= scope['::pulp::max_keep_alive'] %>
+  KeepAlive On
+  MaxKeepAliveRequests <%= scope['::pulp::max_keep_alive'] %>
 
-SSLUsername SSL_CLIENT_S_DN_CN
+<% if scope['::pulp::ssl_username'] -%>
```
Comment (ekohl) [inline templates/etc/httpd/conf.d/_ssl_vhost.conf.erb]: I think it should also check if it's a non-empty string.

File: templates/etc/httpd/conf.d/_ssl_vhost.conf.erb [hunk 2]
```diff
@@ -1,7 +1,9 @@
 # allow older yum clients to connect, see bz 647828
```
Comment (ekohl) [inline templates/etc/httpd/conf.d/_ssl_vhost.conf.erb]: nit: if you're indenting the rest, you should indent this as well.

Comment (llabrat) [inline templates/etc/httpd/conf.d/_ssl_vhost.conf.erb]: Through some black magic that I didn't bother to figure out, it comes out indented correctly as is.

**Comment by ehelms** [comment]
@llabrat in the examples above you showed what the 'CN' of the Pulp certificate looked like -- would you be able to paste me what the 'DN' field shows int he cert?

**Comment by llabrat** [comment]
@ehelms Here is the full output:
```
# openssl x509 -inform pem -in user-cert.pem -noout -text
Certificate:
    Data:
        Version: 1 (0x0)
        Serial Number: 11 (0xb)
    Signature Algorithm: sha1WithRSAEncryption
        Issuer: CN=vagrant, O=PULP
        Validity
            Not Before: Aug 31 14:18:22 2016 GMT
            Not After : Sep  7 14:18:22 2016 GMT
        Subject: CN=admin:admin:57c58e2004c06d1af2f27a8f
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (1024 bit)
                Modulus:
                    00:ca:42:19:59:88:99:fa:11:59:c8:f4:4f:3b:03:
                    27:18:57:21:18:92:bc:d3:34:39:18:2c:09:64:48:
                    b5:1c:cf:c7:21:2f:c0:f1:9e:38:4c:43:9f:aa:e6:
                    fa:eb:a8:bf:1a:5c:5b:99:6a:1d:55:1c:ed:24:21:
                    0c:04:27:b5:4d:31:25:85:c6:22:e8:d8:94:9e:26:
                    ef:6d:2c:65:57:a9:65:f3:8a:7b:24:51:0f:f1:c4:
                    10:bc:6e:98:4a:20:53:8f:de:b3:38:3f:af:b5:4e:
                    7c:0f:3d:a1:f0:92:3e:79:79:82:17:2c:5a:a0:51:
                    15:13:3f:19:e5:e8:38:ed:63
                Exponent: 65537 (0x10001)
    Signature Algorithm: sha1WithRSAEncryption
         72:02:37:e6:b9:ac:59:f3:f2:6f:91:d4:59:d8:00:ef:1f:3c:
         5b:ab:f0:ca:81:b1:4e:65:0c:2e:79:b6:07:1e:a2:21:ce:75:
         c7:c7:d0:b0:5c:4f:3e:9b:de:30:cc:df:04:57:52:fe:90:51:
         bc:68:4f:32:f7:4a:26:75:96:c3:be:ae:98:5c:78:2c:c3:db:
         58:6d:49:67:43:2a:ab:06:ca:8b:01:a5:d4:84:c3:24:2b:d3:
         37:a9:60:44:d0:9a:56:c7:ae:50:03:a7:37:40:9a:1f:b2:94:
         64:0a:db:ac:f9:0e:d6:b2:f8:8d:c3:5a:a1:40:07:6a:38:0b:
         8c:a4:f5:12:8f:59:3e:35:d7:69:10:66:31:d0:66:4c:ce:9b:
         71:fc:5a:bd:4c:f4:f0:5b:6e:54:1f:41:53:1b:da:57:c7:b0:
         04:e1:d2:14:20:9a:77:e9:35:7f:11:bb:17:27:3b:5d:64:d6:
         8f:8d:4b:bd:a5:ed:22:7d:84:54:09:7b:f2:c6:5e:4c:6f:34:
         f2:8e:fe:6b:9c:e0:e7:a9:21:84:d7:ba:48:8a:e3:56:ae:b9:
         54:91:65:a0:e3:56:b1:26:69:85:ee:97:41:6e:c3:13:2c:5f:
         e8:f5:bd:ef:92:14:6e:55:34:0b:85:d7:1c:ff:3c:67:2d:32:
         bb:d7:9d:31:35:36:08:31:43:71:83:06:c5:62:2d:b5:5b:91:
         22:3e:c4:6b:26:7a:1e:77:54:6a:f5:ab:5a:cc:6d:95:cf:bf:
         c6:5c:1e:ac:7a:6d:82:63:08:b8:23:0e:35:03:be:07:fe:c9:
         43:b4:5a:09:5f:0f:24:b0:b4:9e:fb:dc:ea:44:6f:86:56:57:
         51:d3:c5:f4:8d:c9:69:db:ab:44:a5:81:5e:b1:4a:3b:87:0e:
         0a:99:e0:91:83:82:1e:9a:15:90:f7:06:79:3d:a3:69:c0:14:
         d1:b1:2d:7e:aa:53:dc:6b:db:5d:21:41:b7:4f:45:a3:fc:50:
         e4:c6:e1:e8:1a:b4:7d:a6:4b:08:9f:f6:18:3a:be:79:6a:12:
         84:5c:1a:af:ba:6e:80:fd:24:50:60:27:6a:1b:cd:9f:57:32:
         00:71:82:b7:88:61:2e:cd:5f:de:5a:5b:65:94:7a:f7:24:2a:
         87:ca:47:7d:21:bf:ae:5b:59:3d:93:82:20:18:a6:44:dc:10:
         bc:41:e5:a2:2f:83:af:e2:18:4c:8d:97:2f:c9:02:d4:9a:0f:
         5c:58:0c:be:69:a0:f6:3b:06:54:bd:77:9b:73:61:81:59:3f:
         7a:09:fa:12:76:29:aa:86:1e:9d:c4:b6:22:ec:cc:22:dc:24:
         3f:69:47:29:6a:22:71:70
```

**Comment by ehelms** [comment]
Here is what one of our certificates looks like:

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 13130285660916697705 (0xb638253dca584a69)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: C=US, ST=North Carolina, L=Raleigh, O=Katello, OU=SomeOrgUnit, CN=puppet.example.com
        Validity
            Not Before: Aug 25 12:49:48 2016 GMT
            Not After : Aug 27 12:49:49 2036 GMT
        Subject: C=US, ST=North Carolina, O=PULP, OU=NODES, CN=admin
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:a1:26:6e:f5:c1:11:25:e1:d8:90:2e:a9:a1:dd:
                    73:f9:12:64:96:ed:d1:e1:5e:f3:02:d9:dc:79:c3:
                    36:67:e4:9b:9a:44:96:15:57:94:46:0e:24:4e:6f:
                    8e:c9:76:7a:cf:f4:73:ad:ff:db:b5:69:cd:40:b8:
                    e7:e3:06:4d:92:26:87:0b:85:57:cb:8c:9c:63:bf:
                    1e:48:67:9d:cc:4b:d9:f2:17:e8:90:03:d2:0a:2a:
                    38:e9:a9:24:46:ef:5a:64:0a:ea:cc:3c:bb:a4:2c:
                    a5:34:e5:b6:0d:cf:63:78:c0:d3:02:54:f6:5e:cd:
                    b6:e4:1e:0c:b1:e3:e1:20:e9:71:5f:58:38:e8:be:
                    2f:6a:7a:be:d3:30:46:b3:0c:fe:bf:66:52:2d:c5:
                    11:68:87:01:ac:75:1a:21:f2:93:d2:03:fa:33:b6:
                    a7:4e:d0:b6:44:4c:13:7f:13:35:a9:a4:80:0e:c4:
                    51:7a:85:ce:68:f3:34:24:d9:04:af:b2:75:b4:a3:
                    0d:f2:15:6e:e8:8e:fd:dd:93:e8:c1:dd:8d:c0:b7:
                    5d:5a:52:02:25:c5:88:d8:af:f2:eb:49:4d:a3:21:
                    7c:c9:f8:f8:94:87:c7:80:ed:48:a4:fb:9d:e6:31:
                    8b:2a:67:d1:68:80:ea:76:93:22:18:ce:24:4c:22:
                    0a:2b
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: 
                CA:FALSE
            X509v3 Key Usage: 
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication, TLS Web Client Authentication
            Netscape Cert Type: 
                SSL Client
            Netscape Comment: 
                Katello SSL Tool Generated Certificate
            X509v3 Subject Key Identifier: 
                10:AD:F1:A0:26:38:64:39:F8:53:0C:7B:88:E8:01:F9:28:F9:7E:67
            X509v3 Authority Key Identifier: 
                keyid:89:20:96:5F:3C:E6:42:59:EB:00:46:FC:61:8B:0C:12:60:E1:31:81
                DirName:/C=US/ST=North Carolina/L=Raleigh/O=Katello/OU=SomeOrgUnit/CN=puppet.example.com
                serial:B6:38:25:3D:CA:58:4A:66

            X509v3 Subject Alternative Name: 
                DNS:admin
    Signature Algorithm: sha256WithRSAEncryption
         19:a5:22:7b:99:62:cf:fc:c0:db:18:b5:0c:a1:4f:b0:42:fb:
         f5:90:59:a5:bd:81:0e:32:e4:6b:5e:57:a3:2d:85:0c:f6:0c:
         5e:f8:7d:6d:06:0c:48:73:1e:ca:69:33:2c:93:3a:42:68:c5:
         cb:eb:4c:db:c2:ad:33:48:85:6e:69:b5:e8:7b:88:96:1f:1d:
         e1:fe:e9:ec:1b:b6:74:e3:59:b8:76:04:74:20:27:4c:66:5d:
         19:1d:b2:89:e7:ea:4e:af:f2:55:1d:e6:3d:c1:f2:2b:9f:56:
         a3:c3:15:ac:21:2e:4c:06:ac:2c:fb:e4:a5:d8:da:73:c2:b9:
         c7:de:74:22:a5:fa:4b:53:a7:b9:2d:b1:4c:e5:61:6d:f1:ad:
         df:28:9a:00:90:16:7b:b2:2e:f8:c2:05:5e:cd:7f:4f:83:37:
         e2:6b:a9:c7:1a:20:3a:b7:29:fa:12:5a:99:30:02:6e:84:43:
         82:81:e0:d8:9a:9a:37:ed:16:ab:3c:29:52:fd:02:2a:a8:d3:
         18:77:67:83:c7:97:92:40:f1:ee:2f:38:66:ee:68:ce:59:64:
         97:15:c5:7f:42:d3:7c:9c:a4:5c:44:4c:a7:49:bd:24:c5:7c:
         00:de:13:3d:a4:e7:95:cc:ba:11:6c:b4:ae:3f:93:c9:e4:90:
         ea:ea:9d:4b
```

**Comment by ehelms** [comment]
After some digging, config option is the only way to go until Pulp 3.0. Basically, we set SSLUsername to force Pulp to use HTTP basic user authentication by having it set REMOTE_USER. Whereas Pulp admin relies on creating a certificate that embeds username and user.id and goes through a different auth and authorization method.

**Comment by jlsherrill** [comment]
This looks good to me,  @ehelms any reason to not accept?

**Comment by ehelms** [comment]
The default should probably be that this is not set since that would be "normal" working Pulp. Which require us to do some updates to puppet-katello and maybe puppet-capsule. However, I am OK with this as is, I just wanted to point out that fact if looking at this purely from Pulp perspective.

**Comment by ekohl** [comment]
While I agree with @ehelms that we should provide a vanilla setup, right now the module already hardcodes this value. IMHO this should be merged as it now and when we do a major version bump then we should consider changing the default.

**Comment by jlsherrill** [comment]
@ekohl agreed, that was my concern as well

**Comment by ehelms** [comment]
ACK - thanks @llabrat !

**Pull request merged**